packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+        """
+        with self.lock:
+            if self.stopped:
+                self.stopped = False
+                # self.thread.start()
+                # Initializing the internal HTTP server
+                try:
+                    # self.httpd = HTTPServer((self.host, self.port), HTTPHandler(self))
+                    # handler = SimpleHTTPServer.SimpleHTTPRequestHandler
+                    # handler = HTTPHandler.set_channel(self)
+                    # handler.set_channel(self)
+                    self.httpd = HTTPServer(
+                        (self.host, self.port), HTTPHandlerFactory(self)
+                    )
+                    # self.httpd = HTTPServer((self.host, self.port), handler)
+                    logger.info(f"HTTP Server has connected to port {self.port}.")
+                    self.httpd.serve_forever()
+                except OSError:
+                    logger.error(
+                        f"{self.host}:{self.port} is already in use, please try another Socket."
+                    )
+
+    def disconnect(self) -> None:
+        """
+        Disconnect.
+
+        Shut-off the HTTP Server and then HTTP Channel.
+        """
+        with self.lock:
+            if not self.stopped:
+                self.httpd.shutdown()
+                logger.info(f"HTTP Server has shutdown on port {self.port}.")
+                self.stopped = True
+                # self.thread.join()
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
+    class HTTPHandler(BaseHTTPRequestHandler):
+        def __init__(self, *args, **kwargs):
+            self._channel = channel
+            super(HTTPHandler, self).__init__(*args, **kwargs)
+
+        def do_HEAD(self):
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+        def do_GET(self):
+            """Respond to a GET request."""
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+            parsed_path = urlparse(self.path)
+            url = parsed_path.path
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
+            self._channel._process_request("GET", url, param)
+
+        def do_POST(self):
+            """Respond to a POST request."""
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+            parsed_path = urlparse(self.path)
+            url = parsed_path.path
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
+            content_length = int(self.headers["Content-Length"])
+            body = self.rfile.read(content_length).decode()
+
+            self._channel._process_request("POST", url, param, body)
```

Comment:
same as for get request, I'd wait for a response to determine the status code and then serve the status code https://pymotw.com/2/BaseHTTPServer/

So `response = self._channel._process_request("POST", url, param, body)` and then serve it.

Reply:
Same thing as GET. Although now both calls the function `_send` in **HTTPHandlerFactory** that takes care of the Envelope and response so that `do_GET` and `do_POST` don't have to individually.